lib/metrics.js

```diff
@@ -22,169 +14,56 @@ class MetricsEvent {
     this.totalJobsCompleted = totalJobsCompleted
     this.totalEarnings = totalEarnings
   }
-
-  /**
-   * @param {String} line
-   * @returns {MetricsEvent}
-   */
-  static fromLogLine (line) {
-    return new MetricsEvent(JSON.parse(parseLog(line).text))
-  }
 }
 
-class Metrics {
-  constructor (metricsPath, allMetricsPath, logs) {
-    this.metricsPath = metricsPath
-    this.allMetricsPath = allMetricsPath
-    this.logs = logs
-  }
-
-  #getMetricsFilePath (module) {
-    return module
-      ? join(this.metricsPath, `${module}.log`)
-      : this.allMetricsPath
-  }
-
-  /**
-   * @param {string=} module
-   * @returns {Promise<MetricsEvent>}
-   */
-  async getLatest (module) {
-    const metrics = await fs.readFile(this.#getMetricsFilePath(module), 'utf-8')
-    return MetricsEvent.fromLogLine(metrics.trim().split('\n').pop())
-  }
-
-  /**
-   * @param {Object} [options]
-   * @param {String} [options.module]
-   * @param {AbortSignal} [options.signal]
-   * @returns {AsyncGenerator<MetricsEvent>}
-   */
-  async * follow ({ module, signal } = {}) {
-    const tail = new Tail(this.#getMetricsFilePath(module), {
-      nLines: 1,
-      useWatchFile: platform() === 'win32'
-    })
-    for await (const [line] of on(tail, 'line', { signal })) {
-      yield MetricsEvent.fromLogLine(line)
-    }
+class Metrics extends EventEmitter {
+  constructor () {
+    super()
+    this.mergedMetrics = null
+    this.moduleMetrics = new Map()
   }
```

Comment:
I tried making `emit` uncallable:

```js
emit () {
  throw new Error('Use Metrics#submit')
}
```

Typescript complained that the method doesn't adhere to the required return type of `Boolean`. If I add a `return true` after the `throw`, the linter wants to remove it.

Reply:
I think TypeScript can understand that your method never returns, therefore it infers the return type as `never` ([docs](https://www.typescriptlang.org/docs/handbook/basic-types.html#never)).

Would the following alternative work?

```js
/** @returns {Boolean} */
emit () {
  throw new Error('Use Metrics#submit')
}
```

Reply:
Alternatively, maybe we should not use inheritance. If we don't want Metrics to act as an event emitter, then we should not inherit from it.

```js
class Metrics {
  #emitter = new EventEmitter();
  // ...
  submit (moduleName, metrics) {
    // ...
    this.#emitter.emit(/*...*/);
  }

  onChange(fn) {
    this.#emitter.on('update', (metrics) => fn(metrics));
  }
}
```

This will also solve the problem of providing TypeScript typings for events. The event emitter is an internal implementation detail; thus it's kind of ok to have no types for it.

Reply:
This helped fix a bug right away, done :)